src/shared/machine-credential.c

```diff
@@ -0,0 +1,139 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "alloc-util.h"
+#include "creds-util.h"
+#include "escape.h"
+#include "extract-word.h"
+#include "fileio.h"
+#include "macro.h"
+#include "memory-util.h"
+#include "machine-credential.h"
+#include "path-util.h"
+#include "string-util-fundamental.h"
+
+static void machine_credential_done(MachineCredential *cred) {
+        assert(cred);
+
+        cred->id = mfree(cred->id);
+        cred->data = erase_and_free(cred->data);
+        cred->size = 0;
+}
+
+void machine_credential_free_all(MachineCredential *creds, size_t n) {
+        assert(creds || n == 0);
+
+        FOREACH_ARRAY(cred, creds, n)
+                machine_credential_done(cred);
+
+        free(creds);
+}
+
+int machine_credential_set(MachineCredential **credentials, size_t *n_credentials, const char *cred_string) {
+        _cleanup_free_ char *word = NULL, *data = NULL;
+        MachineCredential *creds = *ASSERT_PTR(credentials);
+        ssize_t l;
+        size_t n_creds = *ASSERT_PTR(n_credentials);
+        int r;
+        const char *p = ASSERT_PTR(cred_string);
+
+        if (n_creds > 0)
+                ASSERT_PTR(creds);
```

Comment:
this is weird. in theory at least we allow systemd to be compiled with -DNDEBUG. This means that asserts are suppressed, i.e. `ASSERT_PTR(x)` loses it's power and just becomes `x`. But if this happens with this line, then you still do the if check beforehand, which is kinda pointless, because it conditionalizes a NOP...

hence please do:

```
assert(creds || n_creds == 0);
```

or so, which properly suppresses the whole assert expression if NDEBUG is set.

Reply:
Fixed in 69a7728f8a.